### ISSUE TITLE
[CREATE_COMPONENT_FAILURE] Used too broadly and misleading?

### ISSUE THREAD

**Issue by ryan-singleton** [dapr/dapr#6668]
We set out to create a Dapr infrastructure that uses pluggable components that are managed as docker images for deployment to a team.

Running the pluggables on the Linux host alongside Dapr was quite easy, initially. As soon as we started publishing the pluggables as docker images, we would encounter this error message.
`[CREATE_COMPONENT_FAILURE]`

The verbiage of this message and the ensuing search results that it turns up largely suggest that indeed the component failed to create. So it is not running.

We sunk days into this and probably should have tried to curl the pluggable container earlier, but when we eventually did, we realized it was indeed running fine.

From here, it was rather apparent that it was a difference in communication or networking between Dapr and a Docker container as opposed to a natively hosted one. I'm sure that some of you probably already know what the problem is, but we did not yet. Our next search approach was "Dapr and Docker container networking". As far as I can tell, all of the documentation, samples, and examples of this currently focus on network host mode, port forwarding, etc. This is because of an assumption that the docker container would always contain the Dapr consuming app. Not the Dapr serving app (a pluggable component).

This is another impediment that was misleading and costly before we realized that one simple fact. We spent an inordinate amount of time on this as well.

Had Dapr originally found that an expected component did **not** provide an expected Unix socket (.sock file), and said so, we would have had this problem resolved within minutes. Honestly, had it not misdirected us and led us to believe that it was a pluggable runtime error, the Unix socket might have been fresh enough in our minds that we would've tried it much earlier. But the misdirections really ate our lunch.

I realize that pluggable components are still rather new, but I do hope that the above inverted dependency scenario will eventually be better covered in the docs. As I learn Dapr better, I might even be confident enough to contribute that.

But for the moment, I do believe that much pain and time can be saved for future Dapr customizing adopters if the unhappy paths that yield `[CREATE_COMPONENT_FAILURE]` be re-evaluated and provide messages that keep adopters on task. Thoughts?

---

Edit

For future readers, by the way, the solution was to create a new docker volume (examples in docker compose format)
```yaml
volumes:
  dapr-sockets:
```
and then mount that volume on **both** your dapr run and your pluggable image services, pointing to the directory containing the socket
```yaml
volumes:
      - dapr-sockets:/tmp/dapr-components-sockets:rw
```

But do **NOT** do a direct bind like this, because the socket will be occupied
```yaml
volumes:
      - /tmp/dapr-components-sockets:/tmp/dapr-components-sockets
```
`:rw` will not help here

If you're not using docker compose, keep in mind that my above examples result in a `docker run` that calls `dapr run`, not the usual `dapr run`-only format.

**Comment by tmacam** [comment]
Hi, @ryan-singleton. Sorry for the bad experience. Pluggable components has quite a few rough edges and its always great to hear about ways to improve it.

Just wondering, in the list below, what you think would be the most effective way to avoid this sort of issues for future developers:

1. Improving the error message that comes with `CREATE_COMPONENT_FAILURE`
2. Adding a section for Docker on https://docs.dapr.io/operations/components/pluggable-components-registration/ or on another page.
3. [Improve the message regarding pluggable component discovery, which right now is a DEBUG and not an INFO level.](https://github.com/dapr/dapr/blob/0b22303f016ab0fb0fcfff050034efb559996e4f/pkg/components/pluggable/discovery.go#L135). Also mention the directory in the log message.



For **1** the challenge I foresee is that, by the time we are processing a component spec, there isn't anything, no marker or configuration flag, that allows us to identify something as a pluggable component. [When we see a request for a component we can't create][1], we don't know for sure if it was just a typo, a component that isn't available on a given Dapr version or a Pluggable component that wasn't discovered*.  Perhaps just saying something like `You might have a typo in your component type or dapr couldn't discover the corresponding pluggable component.` could address it but I don't know.

Improving documentation, as mentioned in **2**, might be appropriate regardless but I wonder if this could have been the single thing that could have saved your time. 
As for item **3**, it could help if you happened to search for "pluggable" in the logs. And perhaps is the simplest thing we could do to avoid such issues in the future. It doesn't point you to the Docker shared volume directly but stating that no pluggable component was found in a given path might set you on the right track.

Thoughts?

`*` Pluggable components are [discovered](https://github.com/dapr/dapr/blob/0b22303f016ab0fb0fcfff050034efb559996e4f/pkg/components/pluggable/discovery.go#L79) during initialization and registered. This takes place before we process the components specs for your dapr application.

[1]: https://github.com/dapr/dapr/blob/0b22303f016ab0fb0fcfff050034efb559996e4f/pkg/runtime/processor/state.go#L50-L53

**Comment by ryan-singleton** [comment]
Hello, and thank you for the thorough response.

For what it's worth, I did have the log level set to debug and I did see that it saw my component spec and was making an attempt to register it. But what happened next was where I got stuck in the mire.

The reality is that we devs are human and can get really stuck in an idea. And this may have been a "skill issue" as they say. Lol. I won't be hurt if that's the determination. But if you are open to some proposals, here are some.

[COMPONENT_REGISTRATION_FAILURE] The component specification may be incorrect or Dapr was unable to discover the component.

This would happen in an immediate sense and possibly prevent a developer from starting down the wrong track that I took.

In any case, I can say that if the Dapr documentation covered examples of pluggable components in other docker containers, my wild goose chase would have been short circuited because I can guarantee that I would have found that documentation fairly quickly and I would have re-routed.

So to me, the real resolution to this is probably that. Documentation. I am willing to write a rough draft of that documentation, if you would like. It is significantly fresh in my mind, lol.

Let me know if you are interested, I would be glad to contribute.
